### PR TITLE
Add `--extra` to `uv add` and enable fine grained updates

### PR DESCRIPTION
**Pull request by ibraheemdev**
## Summary

- Adds a `--extra` flag to `uv add` that allows activating extras without the PEP508 syntax.
- `uv add` now errors if the update is ambiguous (e.g. the dependency is present twice with different markers)
- `uv add` is smarter about updates. For example, `uv add flask==3.0.0` followed by `uv add flask --extra dotenv` preserves the previous version specifier.

Resolves https://github.com/astral-sh/uv/issues/4419.

### CODE REVIEW

File: crates/uv-distribution/src/pyproject_mut.rs [hunk 1]
```diff
@@ -194,19 +197,45 @@ fn implicit() -> Item {
...
     }
     reformat_array_multiline(deps);
+    Ok(())
+}
+
+/// Update an existing requirement.
+fn update_requirement(old: &mut Requirement, new: Requirement, has_source: bool) {
+    // Add any new extras.
+    old.extras.extend(new.extras);
+    old.extras.dedup();
```
Comment (charliermarsh) [inline crates/uv-distribution/src/pyproject_mut.rs:221]: Do we need to sort?

Comment (ibraheemdev) [reply crates/uv-distribution/src/pyproject_mut.rs:221]: I don't think so? This is updating `pyproject.toml` not the lockfile.

Comment (charliermarsh) [reply crates/uv-distribution/src/pyproject_mut.rs:221]: But `dedup` only deduplicates adjacent entries, right?

Comment (ibraheemdev) [reply crates/uv-distribution/src/pyproject_mut.rs:222]: Oh right, fixed. I wonder if `extras` should be a `HashSet`..

**Review by charliermarsh — APPROVED** [review]

**Pull request merged**